Block and critical formulas implementation

kenedos opened the pull request:
This PR implements:

- [NEW] Block Rate
- [NEW] Block Penetration
- Block Chance Formula
- Block damage reduction (50% less damage)
- [NEW] Critical Defense Rate
- Improved Critical Chance Formula
- Improved Critical Damage to deal 1.5x damage
- Adds Block/BlockPen and Critical Attack/Rate/Defense stats for monsters

This PR *DOES NOT* implement:

- Active blocking (holding 'C' to block)
- Equipment block rates in items.txt database

Videos:

- Player blocking attacks: https://www.youtube.com/watch?v=tGhd0X7mmpM
- Monster blocking attacks: https://www.youtube.com/watch?v=IcUwUen4OZA

Based on: https://treeofsavior.com/page/news/view.php?n=951
Credits to @SalmanTKhan for base implementation

exectails commented on system/scripts/zone/core/calc_monster.cs:
```diff
@@ -334,4 +334,140 @@ public float SCR_Get_MON_SDR(Mob monster)
+		byBuffs += monster.Properties.GetFloat(PropertyName.CRTATK_BM, 0);
+		
+		var byRateBuffs = 0f;
+		//byRateBuffs += monster.Properties.GetFloat(PropertyName.CRTATK_RATE_BM);
```
This property doesn't appear to exist. Could these lines not be removed then?

exectails commented on system/scripts/zone/core/calc_monster.cs:
```diff
@@ -334,4 +334,140 @@ public float SCR_Get_MON_SDR(Mob monster)
+		var byBuffs = properties.GetFloat(PropertyName.BLK_BM);
+		
+		var byRateBuffs = 0f;
+		byRateBuffs += properties.GetFloat(PropertyName.BLK_RATE_BM);
```
Same as the other one, this property does not seem to exist for monsters.

exectails commented on system/scripts/zone/core/calc_monster.cs:
```diff
@@ -334,4 +334,140 @@ public float SCR_Get_MON_SDR(Mob monster)
+		var byBuffs = properties.GetFloat(PropertyName.BLK_BREAK_BM);
+		
+		var byRateBuffs = 0f;
+		byRateBuffs += properties.GetFloat(PropertyName.BLK_BREAK_RATE_BM);
```
And this one doesn't exist for monsters either. I checked some older data as well, and it doesn't seem like these properties ever existed for monsters. They didn't even exist for players in older versions. I wonder if we should leave them, in case they're ever added, or if we should rather remove them to avoid confusion.

kenedos replied:
Maybe I'll leave them commented for now so we remember they exist but they're not actually implemented?
Or should I just remove them entirely?
What would be the best solution in your opinion?

exectails replied:
I'm one the fence. They don't exist for monsters currently, so including them in any way could lead to confusion. Either because people don't see them anywhere, or, if they're commented, because one might think they just weren't implemented by us yet. I believe it would be best to remove them until they're actually added for monsters, if ever.

kenedos commented:
I've removed the nonexisting monster properties mentioned:
- CRTATK_RATE_BM
- BLK_RATE_BM
- BLK_BREAK_RATE_BM

I've also removed the `(int)Math.Max(0, value)` from return of those monster_calc functions, as it seems to make more sense leaving those able to go negative.

Please let me know if there's anything else needed in regards to this PR.

exectails commented:
Looks good I think. Thanks for the PR.
